pkg/cmd/roachtest/tests/mixed_version_cdc.go

```diff
@@ -0,0 +1,318 @@
+		},
+	})
+}
+
+// cdcMixedVersionTester implements mixed-version/upgrade testing for
+// CDC. It knows how to set up the cluster to run Kafka, monitor
+// events, and run validations that ensure changefeeds work during and
+// after upgrade.
+type cdcMixedVersionTester struct {
+	ctx           context.Context
+	crdbNodes     option.NodeListOption
+	workloadNodes option.NodeListOption
+	kafkaNodes    option.NodeListOption
+	monitor       cluster.Monitor
+	verifierDone  chan struct{}
+	kafka         kafkaManager
+	validator     *cdctest.CountValidator
+	cleanup       func()
+}
+
+func newCDCMixedVersionTester(
+	ctx context.Context, t test.Test, c cluster.Cluster,
+) cdcMixedVersionTester {
+	crdbNodes := c.Range(1, c.Spec().NodeCount-1)
+	lastNode := c.Node(c.Spec().NodeCount)
+
+	c.Put(ctx, t.Cockroach(), "./cockroach", lastNode)
+	c.Put(ctx, t.DeprecatedWorkload(), "./workload", lastNode)
+
+	return cdcMixedVersionTester{
+		ctx:           ctx,
+		crdbNodes:     crdbNodes,
+		workloadNodes: lastNode,
+		kafkaNodes:    lastNode,
+		monitor:       c.NewMonitor(ctx, crdbNodes),
+		verifierDone:  make(chan struct{}),
+	}
+}
+
+// StartKafka will install and start Kafka on the configured node. It
+// will also create the topic where changefeed events will be sent.
+func (cmvt *cdcMixedVersionTester) StartKafka(t test.Test, c cluster.Cluster) {
+	t.Status("starting Kafka node")
+	cmvt.kafka, cmvt.cleanup = setupKafka(cmvt.ctx, t, c, cmvt.kafkaNodes)
+	if err := cmvt.kafka.createTopic(cmvt.ctx, target); err != nil {
+		t.Fatal(err)
+	}
+}
+
+// Cleanup is supposed to be called at the end of tests that use
+// `cdcMixedVersionTester`
+func (cmvt *cdcMixedVersionTester) Cleanup() {
+	if cmvt.cleanup != nil {
+		cmvt.cleanup()
+	}
+}
+
+// installAndStartTPCCWorkload starts a TPCC workload for the given
+// duration. This step does not block; if an error is found while
+// runing the workflow, the test fails.
+func (cmvt *cdcMixedVersionTester) installAndStartTPCCWorkload(d string) versionStep {
+	return func(ctx context.Context, t test.Test, u *versionUpgradeTest) {
+		t.Status("installing and running workload")
+		tpcc := tpccWorkload{
+			sqlNodes:           cmvt.crdbNodes,
+			workloadNodes:      cmvt.workloadNodes,
+			tpccWarehouseCount: 10,
+			tolerateErrors:     true, // we're bringing nodes down while upgrading
+		}
+
+		tpcc.install(ctx, u.c)
+		time.Sleep(2 * time.Second)
+		cmvt.monitor.Go(func(ctx context.Context) error {
+			tpcc.run(ctx, u.c, d)
+			return nil
+		})
+	}
+}
+
+// waitForVerifier waits for the underlying CDC verifier to reach the
+// desired number of resolved timestamps.
+func (cmvt *cdcMixedVersionTester) waitForVerifier() versionStep {
+	return func(ctx context.Context, t test.Test, u *versionUpgradeTest) {
+		t.Status("waiting for verifier")
+		<-cmvt.verifierDone
+	}
+}
+
+// setupVerifier creates a CDC validator to validate that a changefeed
+// created on the `tpcc.warehouse` table is able to re-create the
```

Comment:
Why `warehouse` instead of say `newOrder`? After `tpcc.install`, the warehouse table is never updated whereas `newOrder` is updated with the highest probability. The default weights for each updated table are here: https://github.com/cockroachdb/cockroach/blob/master/pkg/workload/tpcc/tpcc.go#L193

Reply:
Interestingly, I tried using `new_order` as it's the only table that doesn't have a decimal column, following @HonoreDB's suggestion. For some reason, however, the test timed out when I did that: it wouldn't reach 20 resolved timestamps after the allotted 30mins. I didn't spend too much time debugging that, but I'll investigate.

Reply:
Interesting; definitely worth debugging this further. This might be indicative of an actual issue.

Reply:
+1 -- definitely interesting; new_order is also substantially larger and more interesting than warehouse